engine/api/src/main/scala/pl/touk/nussknacker/engine/api/typed/CanBeSubclassDeterminer.scala

```diff
@@ -14,93 +17,153 @@ import pl.touk.nussknacker.engine.api.typed.typing._
           case _ => Map.empty[String, TypingResult]
         }
-        superclass.fields.forall {
-          case (name, typ) => givenTypeFields.get(name).exists(canBeSubclassOf(_, typ))
-        }
+
+        joinValidatedNels(superclass.fields.toList.map {
+          case (name, typ) => givenTypeFields.get(name) match {
+            case None =>
+              f"Field '$name' is lacking".invalidNel
+            case Some(givenFieldType) =>
+              ensureOrInvalid(canBeSubclassOf(givenFieldType, typ).isValid)(
+                f"Field '$name' is of the wrong type. Excepted: ${givenFieldType.display}, actual: ${typ.display}"
+              )
+          }
+        })
       case _ =>
-        true
+        ().validNel
     }
-    def dictRestriction: Boolean = (givenType, superclassCandidate) match {
-      case (given: TypedDict, superclass: TypedDict) =>
-        given.dictId == superclass.dictId
-      case (_: TypedDict, _) =>
-        false
-      case (_, _: TypedDict) =>
-        false
-      case _ =>
-        true
+    val dictRestriction = (_: Unit) => {
+      val errorPrefix = "Dict restriction not met:"
+      (givenType, superclassCandidate) match {
+        case (given: TypedDict, superclass: TypedDict) =>
+          ensureOrInvalid(given.dictId == superclass.dictId)(
+            f"$errorPrefix ${givenType.display} and ${superclassCandidate.display} are dicts with unequal IDs")
+        case (_: TypedDict, _) =>
+          f"$errorPrefix ${givenType.display} is a Dict but ${superclassCandidate.display} not".invalidNel
+        case (_, _: TypedDict) =>
+          f"$errorPrefix ${superclassCandidate.display} is a Dict but ${givenType.display} not".invalidNel
+        case _ =>
+          ().validNel
+      }
     }
-    def taggedValueRestriction: Boolean = (givenType, superclassCandidate) match {
-      case (givenTaggedValue: TypedTaggedValue, superclassTaggedValue: TypedTaggedValue) =>
-        givenTaggedValue.tag == superclassTaggedValue.tag
-      case (_: TypedTaggedValue, _) => true
-      case (_, _: TypedTaggedValue) => false
-      case _ => true
+    val taggedValueRestriction = (_: Unit) => {
+      (givenType, superclassCandidate) match {
+        case (givenTaggedValue: TypedTaggedValue, superclassTaggedValue: TypedTaggedValue) =>
+          ensureOrInvalid(givenTaggedValue.tag == superclassTaggedValue.tag)(
+            f"Tagged values have unequal tags: ${givenTaggedValue.tag} and ${superclassTaggedValue.tag}")
+        case (_: TypedTaggedValue, _) => ().validNel
+        case (_, _: TypedTaggedValue) =>
+          f"The type is not a tagged value".invalidNel
+        case _ => ().validNel
+      }
     }
-    klassCanBeSubclassOf(givenType.objType, superclassCandidate.objType) && typedObjectRestrictions && dictRestriction &&
+    klassCanBeSubclassOf(givenType.objType, superclassCandidate.objType) andThen typedObjectRestrictions andThen dictRestriction andThen
```

Comment:
andThen or sequence? First probably should be andThen, but the rest?

Reply:
You are right. All the restrictions are desired to be displayed to a user; e.g. there could be a Dict with an Object value type and tagged in another Dict.
So:
```
klassCanBeSubclassOf(givenType.objType, superclassCandidate.objType) andThen
(typedObjectRestrictions combine dictRestriction combine taggedValueRestriction)
```